File: Algorithm/DP/1463.swift
import Foundation

let X = Int(readLine()!)!

print(solution(X))

func solution(_ X: Int) -> Int {
    if X == 1 { return 0 } 
    
    var dp = Array(repeating: 0, count: X + 1)

    if X >= 2 {
        dp[2] = 1
    }
    
    if X >= 3 {
        dp[3] = 1
    }

    for i in 2...X {
        dp[i] = dp[i - 1] + 1

        if i % 2 == 0 {
            dp[i] = min(dp[i], dp[i / 2] + 1)
        }
        
        if i % 3 == 0 {
            dp[i] = min(dp[i], dp[i / 3] + 1)
        }
    }
    
    return dp[X]
}

File: Algorithm/DP/1463.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int INF = 1e9;

int solution(int n) {
    vector<int> dp(n + 1, INF);
    dp[1] = 0;

    if (n >= 2) {
        dp[2] = 1;
    }

    if (n >= 3) {
        dp[3] = 1;
    }

    for (int i = 2; i <= n; ++i) {
        dp[i] = dp[i - 1] + 1;
        if (i % 2 == 0) {
            dp[i] = min(dp[i] , dp[i / 2] + 1);
        }
        if (i % 3 == 0) {
            dp[i] = min(dp[i] , dp[i / 3] + 1);
        }
    }

    // for (int i = 1; i <= n; ++i) {
    //     cout << dp[i] << ' ';
    // }
    // cout << '\n';

    return dp[n];
}

int main(void) {
    int n;

    cin >> n;
    cout << solution(n) << '\n';

    return 0;
}